### PR TITLE
Support for SAML2 artifact binding

### PR DESCRIPTION
**Pull request by vihanga-liyanage**
### Proposed changes in this pull request
- IS Samples can now accept SAML2 artifact responses from an IDP.
- Implementations done in PR https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/63 has been followed with slight modifications. 

Refers wso2/product-is#3401
### When should this PR be merged
ASAP

### Follow up actions
- None

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.agent/pom.xml [hunk 1]
```diff
@@ -82,6 +82,11 @@
             <groupId>org.wso2.orbit.org.owasp.encoder</groupId>
             <artifactId>encoder</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.3.1.wso2v1</version>
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/pom.xml]: Bring this version to root pom

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 2]
```diff
@@ -365,6 +368,77 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+     * @throws SSOAgentException
+     */
+    private void executeSAMLResponse(HttpServletRequest request, XMLObject samlObject) throws SSOAgentException {
+        if (samlObject instanceof LogoutResponse) {
+            //This is a SAML response for a single logout request from the SP
+            doSLO(request);
+        } else if (samlObject instanceof Response) {
+            processSSOResponse(request, (Response) samlObject);
+        } else {
+            throw new SSOAgentException("Unable to process unknown SAML object type.");
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: Isn't that a good idea to add object type as well ?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 3]
```diff
@@ -365,6 +368,77 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+    /**
+     * Redirect SAML response received via artifact resolution.
+     *
+     * @param request    Initial authentication request.
+     * @param samlObject Response received via artifact resolution.
+     * @throws SSOAgentException
+     */
+    private void executeSAMLResponse(HttpServletRequest request, XMLObject samlObject) throws SSOAgentException {
+        if (samlObject instanceof LogoutResponse) {
+            //This is a SAML response for a single logout request from the SP
```
Comment (JKAUSHALYA) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: Please follow same convention for all comments.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java [hunk 4]
```diff
@@ -273,6 +297,56 @@ public static void sendPostResponse(HttpServletRequest request, HttpServletRespo
...
+            ByteArrayOutputStream byteArrayOutputStrm = new ByteArrayOutputStream();
+            DOMImplementationRegistry registry = DOMImplementationRegistry.newInstance();
+            DOMImplementationLS impl = (DOMImplementationLS) registry.getDOMImplementation("LS");
+            LSSerializer writer = impl.createLSSerializer();
+            LSOutput output = impl.createLSOutput();
+            output.setByteStream(byteArrayOutputStrm);
+            writer.write(element, output);
+            return byteArrayOutputStrm.toString();
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "Error Serializing the SAML Response");
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java]: Why do we need to do both logging and throwing ?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java [hunk 5]
```diff
@@ -368,4 +442,16 @@ private static Document getDocument(DocumentBuilderFactory documentBuilderFactor
         ByteArrayInputStream inputStream = new ByteArrayInputStream(samlString.getBytes());
         return docBuilder.parse(inputStream);
     }
+
+    public static boolean isArtifactResolveReqSigningEnabled(Map<String, String> properties) {
+
+        if (properties != null) {
+            String prop =
+                    properties.get(SSOAgentConstants.SSOAgentConfig.SAML2.IS_ARTIFACT_RESOLVE_REQ_SIGNED);
+            if (StringUtils.isNotBlank((prop))) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java]: Do we really need this check ?. Isn't Boolean. parseBoolean handle empty values ?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java [hunk 6]
```diff
@@ -273,6 +297,56 @@ public static void sendPostResponse(HttpServletRequest request, HttpServletRespo
...
+     * Serializing a SAML2 object into a String
+     *
+     * @param xmlObject object that needs to serialized.
+     * @return serialized object
+     * @throws SSOAgentException
+     */
+    public static String marshall(XMLObject xmlObject) throws SSOAgentException {
+        try {
+
+            System.setProperty("javax.xml.parsers.DocumentBuilderFactory",
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/util/SSOAgentUtils.java]: Why do we need to set this system property explicitly ? If we are forcefully setting this, don't we need to remove this as well ?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        }
+
+        String artifactResponseString = soapMessageService.sendSOAP(envelopeElement,
+                ssoAgentConfig.getSAML2().getArtifactResolveURL());
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
+        if (artifactResponse != null && isArtifactResponseValid(artifactResolve, artifactResponse)) {
+            return artifactResponse;
+        } else {
+            throw new ArtifactResolutionException("Artifact Response is not valid.");
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: Don't we miss some information here ?. When this exception is thrown to upper level we don't have a clue on what actually the response was. Isn't that a good idea to log the response somewhere ?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 8]
```diff
@@ -365,6 +368,77 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+    public void processArtifactResponse(HttpServletRequest request) throws SSOAgentException {
+
+        SAMLSSOArtifactResolutionService artifactResolutionService =
+                new SAMLSSOArtifactResolutionService(ssoAgentConfig);
+        try {
+            ArtifactResponse artifactResponse = artifactResolutionService.getSAMLArtifactResponse(
+                    request.getParameter(SSOAgentConstants.SAML2SSO.SAML2_ARTIFACT_RESP));
+
+            if (!StringUtils.equals(artifactResponse.getStatus().getStatusCode().getValue(), StatusCode.SUCCESS_URI)) {
+                throw new SSOAgentException("Received an invalid SAML response.");
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: Add the status value to the error.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                    StringWriter stringResult = new StringWriter();
+                    TransformerFactory.newInstance().newTransformer().transform(
+                            source, new StreamResult(stringResult));
+                    artifactResponse = (ArtifactResponse) SSOAgentUtils.unmarshall(stringResult.toString());
+                } else {
+                    throw new ArtifactResolutionException("Didn't receive valid artifact response.");
+                }
+            }
+        } catch (SOAPException | IOException | TransformerException e) {
+            throw new ArtifactResolutionException("Didn't receive valid artifact response.", e);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: Add more information to error message.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 10]
```diff
@@ -365,6 +368,77 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+            }
+            if (artifactResponse.getMessage() == null) {
+                throw new SSOAgentException("Received SAML2 Artifact response message was null.");
+            }
+            XMLObject xmlObject = artifactResponse.getMessage();
+            if (xmlObject instanceof Response || xmlObject instanceof LogoutResponse) {
+                validateArtifactResponseMessage(xmlObject);
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: let's differentiate artifact response and actual saml response.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 11]
```diff
@@ -365,6 +368,77 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+                    request.getParameter(SSOAgentConstants.SAML2SSO.SAML2_ARTIFACT_RESP));
+
+            if (!StringUtils.equals(artifactResponse.getStatus().getStatusCode().getValue(), StatusCode.SUCCESS_URI)) {
+                throw new SSOAgentException("Received an invalid SAML response.");
+            }
+            if (artifactResponse.getMessage() == null) {
+                throw new SSOAgentException("Received SAML2 Artifact response message was null.");
+            }
+            XMLObject xmlObject = artifactResponse.getMessage();
+            if (xmlObject instanceof Response || xmlObject instanceof LogoutResponse) {
```
Comment (hasinthaindrajee) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: We can't silently return in a case if this condition is not met

Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: fixed.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/bean/SSOAgentConfig.java [hunk 12]
```diff
@@ -742,6 +745,29 @@ public void checkServerTrusted(java.security.cert.X509Certificate[] certs,
         private String relayState = null;
         private String signatureValidatorImplClass = null;
         private int timeStampSkewInSeconds = 300;
+        private String artifactResolveURL = null;
+        private boolean enableArtifactResolveSigning;
+
+        public void setArtifactResolveURL(String artifactResolveURL) {
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/bean/SSOAgentConfig.java]: Add java doc comments for all the public methods.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java [hunk 13]
```diff
@@ -365,6 +368,81 @@ public void processResponse(HttpServletRequest request, HttpServletResponse resp
...
+     *
+     * @param request Http Servlet Request object.
+     * @throws SSOAgentException
+     */
+    public void processArtifactResponse(HttpServletRequest request) throws SSOAgentException {
+
+        SAMLSSOArtifactResolutionService artifactResolutionService =
+                new SAMLSSOArtifactResolutionService(ssoAgentConfig);
+        try {
+            ArtifactResponse artifactResponse = artifactResolutionService.getSAMLArtifactResponse(
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: Would be better to add debug log to print artifact response.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    public ArtifactResolve generateArtifactResolveReq(String samlArtReceived) throws ArtifactResolutionException {
+
+        ArtifactResolve artifactResolve = createArtifactResolveObject(samlArtReceived);
+        if (ssoAgentConfig.getSAML2().isEnableArtifactResolveSigning()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Signing artifact resolve request for the received SAML artifact: " + samlArtReceived);
+            }
+            artifactResolve = signArtifactResolveReq(artifactResolve);
+        }
+        return artifactResolve;
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: Would be better to include debug logs to print samlArtReceived and created ArtifactResolve request.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+        } catch (SSOAgentException e) {
+            throw new ArtifactResolutionException("Encountered error marshalling SOAP message with artifact " +
+                    "resolve, into its DOM representation", e);
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Artifact Resolve Request as a SOAP Message: " + envelopeElement);
+        }
+
+        String artifactResponseString = soapMessageService.sendSOAP(envelopeElement,
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: Would be better to include debug log to print artifactResponseString.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+                ssoAgentConfig.getSAML2().getArtifactResolveURL());
+        ArtifactResponse artifactResponse = extractArtifactResponse(artifactResponseString);
+
+        validateArtifactResponse(artifactResolve, artifactResponse);
+
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: Remove unnecessary code lines.

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOSoapMessageService.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+     * Send SOAP message
+     *
+     * @param message message that needs to be send
+     * @param url     url that the artifact resolve request should be sent
+     * @return response of invoking artifact resolve endpoint
+     * @throws ArtifactResolutionException
+     */
+    public String sendSOAP(String message, String url) throws ArtifactResolutionException {
+
+        StringBuilder soapResponse = new StringBuilder();
```
Comment (IndunilRathnayake) [inline components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOSoapMessageService.java]: Validate whether message and url is not null. would be better to print those in a debug log.

**Review by IndunilRathnayake — CHANGES_REQUESTED** [review]

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java (hunk 14 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: ArtifactResolve is printed in [here](https://github.com/wso2-extensions/identity-agent-sso/pull/46/files#diff-e58b1bc8dd5cd23b7eb0aa3b3e5d0a68R129) along with SOAP envelope. Should we print it here as well?

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java (hunk 13 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/SAML2SSOManager.java]: Printing artifact response in [here](https://github.com/wso2-extensions/identity-agent-sso/pull/46/files#diff-e58b1bc8dd5cd23b7eb0aa3b3e5d0a68R175).

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java (hunk 15 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/saml/artifact/SAMLSSOArtifactResolutionService.java]: added

File: components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/bean/SSOAgentConfig.java (hunk 12 above)
Comment (vihanga-liyanage) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/bean/SSOAgentConfig.java]: There are simple getters and setters. Should we have doc comments to these as well?

Comment (IndunilRathnayake) [reply components/org.wso2.carbon.identity.sso.agent/src/main/java/org/wso2/carbon/identity/sso/agent/bean/SSOAgentConfig.java]: Yes, these are public methods.

**Review by IndunilRathnayake — APPROVED** [review]

**Comment by jofisiva** [comment]
I am using the OPEN SAML -3.2.0  version will SAML2SSOManager  will run? plan to integrate the SAML2SSOMANAGER with spring framework

**Comment by jofisiva** [comment]
import org.opensaml.xml.util.Base64;
import org.opensaml.xml.util.XMLHelper; 

Base 64 class won't present in the opensaml 3.2.0 jar.I believe SAMLSSOMANAGER is tightly coupled with opensaml 2.6.4

**Pull request merged**